[DR-3034] Delete policies in snapshot delete flight

samanehsan opened the pull request:
https://broadworkbench.atlassian.net/browse/DR-3034

In `SnapshotDeleteFlight`, add a new step `DeleteSnapshotPolicyStep` that deletes the snapshot's policy from the Terra Policy Service if its parent dataset has `secureMonitoringEnabled`. TPS handles the case where the policy is still linked to a Terra workspace, so TDR does not need to perform that check before calling `policyService.deletePao`.

samanehsan committed:
Delete policy from TPS when deleting snapshot

samanehsan committed:
Add unit tests for DeleteSnapshotPolicyStep

samanehsan committed:
Test TPS not enabled when creating policies

For consistency, also add a test for the
CreateSnapshotPolicyStep that verifies the step
will not fail if the policy service is not yet
enabled.

samanehsan committed:
Only call deletePao for snapshots using secure monitoring

samanehsan commented:
The connected test failures are happening in the test clean-up when snapshots are getting deleted, so I'm currently fixing that 🛠️

samanehsan committed:
Fix deleting snapshots that do not exist

If a snapshot does not exist, we assume that
its policy was already deleted along with the
snapshot. By wrapping the step with PerformDatasetStep,
we only call DeleteSnapshotPolicyStep if the source
dataset exists (since we need to check that the source
dataset has secure monitoring enabled).

samanehsan commented:
@snf2ye @okotsopoulos I pushed up a fix for the tests! I needed to handle the case where a snapshot that does not exist is being deleted :octocat:

okotsopoulos commented on src/main/java/bio/terra/service/snapshot/flight/delete/DeleteSnapshotPolicyStep.java:
```diff
@@ -0,0 +1,45 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class DeleteSnapshotPolicyStep extends DefaultUndoStep {
```
Are we sure that this step should have the default undo behavior (that is, return a success on undo)?  I would have thought that undoing the deletion of a snapshot's policy would be to restore that policy.

samanehsan replied:
Oh now that I think of it, that behavior makes sense to me. We wouldn't want the case where a snapshot still exists (because the delete failed) but its policy is gone.

okotsopoulos commented on src/main/java/bio/terra/service/snapshot/flight/delete/DeleteSnapshotPolicyStep.java:
```diff
@@ -0,0 +1,45 @@
+    Dataset dataset = datasetService.retrieve(datasetId);
+    if (dataset.isSecureMonitoringEnabled()) {
+      try {
+        policyService.deletePao(snapshotId);
```
What's the expected behavior if the snapshot in question has no associated policy?  One place this could happen is if we're deleting a snapshot that was created prior to our integration with TPS.

samanehsan replied:
If a policy isn't found when deleting a snapshot, we ignore the exception: https://github.com/DataBiosphere/jade-data-repo/blob/develop/src/main/java/bio/terra/service/policy/PolicyService.java#L77

okotsopoulos replied:
Got it.  Since this behavior wasn't clear to me without reading the source code, what would you think about adding a docstring to the method and possibly renaming it to something like `deletePaoIfExists`?

okotsopoulos commented on src/main/java/bio/terra/service/snapshot/flight/delete/DeleteSnapshotPolicyStep.java:
```diff
@@ -0,0 +1,45 @@
+    if (dataset.isSecureMonitoringEnabled()) {
+      try {
+        policyService.deletePao(snapshotId);
+      } catch (FeatureNotImplementedException ex) {
```
Ah! Seeing this `FeatureNotImplementedException` handling made me think…

I know that we call `policyService.deletePdao` in the undo method for `CreateSnapshotPolicyStep`.  Should we have that handling there as well, so that if a snapshot creation flight gets undone it doesn't fail dismally?

samanehsan replied:
Oh you're totally right! I'll add that in

okotsopoulos replied:
Thanks for adding that in!

Now that I'm seeing the code for both `CreateSnapshotPolicyStep` and `DeleteSnapshotPolicyStep` I am struck at how similar they are.  I could be missing something, but it looks like one step's do is the other's undo, and vice-versa.  What do you think about deriving both steps from a common class to facilitate code sharing and ensure common error handling?

A higher-level musing on errors -- I am struggling to see the value of the `TpsService` methods throwing `FeatureNotImplementedException`s if we're going to catch them everywhere.  It seems like it makes for more code clean-up down the line when we remove the feature flag, where instead we could maybe exit early from those methods with a log but not throw.  Happy to be convinced otherwise though!

samanehsan replied:
I like your suggestion to consolidate the code! Especially since we're planning to add more logic into these steps for other policy types so I could see this getting out of sync by mistake.

For the `FeatureNotImplementedException`s, I missed a spot where I should have caught that exception so I think that proves your point 😅 I was thinking it would result in a lot of repetition in the `TpsService` methods, but we're planning to remove the logic anyways and it will all be in the same class.

okotsopoulos replied:
Awesome! It's nice to be able to take advantage of the symmetry between creating and deleting snapshots here.

okotsopoulos commented on src/test/java/bio/terra/service/policy/flight/CreateSnapshotPolicyStepTest.java:
```diff
@@ -75,4 +76,20 @@ void testNoPolicyDoUndo() throws Exception {
+    mockFlightMap();
+    CreateSnapshotPolicyStep step = new CreateSnapshotPolicyStep(policyService, true);
+    var exception = new FeatureNotImplementedException("Policy service is not enabled");
+    doThrow(exception).when(policyService).createSnapshotPao(SNAPSHOT_ID, policies);
```
If this exception is thrown on `createSnapshotPao`, I think it would also be thrown on `deletePao`.

samanehsan replied:
~It does! The exception gets thrown by `policyServiceConfiguration.tpsEnabledCheck()` in both `createPao` and `deletePao`: https://github.com/DataBiosphere/jade-data-repo/blob/develop/src/main/java/bio/terra/service/policy/PolicyService.java#L70~

After reading your related comment, I see you were referring to the test here! I'll change that as well

okotsopoulos reviewed (commented):
I left a few questions, please feel free to reach out if you'd like to chat!

samanehsan committed:
Handle FeatureNotImplementedException in CreateSnapshotPolicyStep undo

samanehsan committed:
Re-create policy if snapshot delete fails

samanehsan committed:
Rename deletePao and add docstring

samanehsan committed:
Refactor feature-enabled check

Instead of throwing a FeatureNotImplementedException from
the PolicyService methods and catching the exception where
they get called, have the methods return early if TPS is
not enabled.

samanehsan committed:
Consolidate snapshot policy create/delete steps

okotsopoulos commented on src/test/java/bio/terra/service/policy/PolicyServiceTest.java:
```diff
@@ -76,10 +81,8 @@ void testGetProtectedDataPolicyInput() {
 
   @Test
   void testCreateSnapshotDao() throws Exception {
```
Naming nit - I think it's possible that the tests in this class are misnamed! We want `Dao`s to be `Pao`s :)

samanehsan replied:
Oh no! 🤦‍♀️ Thanks for catching that!

okotsopoulos reviewed (approved):
Thanks for addressing my feedback, it looks great!

samanehsan committed:
Fix typo
